yext/answers
# replace PersistentStorage with Storage

This commit removes the old PersistentStorage in favor of our new Storage.
Additionally, it makes FilterSearch and static FilterBox work with backnavigation
and page refresh. It also fixes a bug where hitting backnav with FilterBox (that was
configured with searchOnChange=true) would trigger one extra liveapi verticalsearch
per FilterOptions inside the FilterBox.

To replace PersistentStorage, a few steps were taken.
1. replace all usages of persistentStorage.set() and delete() with storage.setWithPersist() and delete()
2. replace usages of window.location.search.substr(1) with storage.getCurrentStateUrlMerged() where needed
3. call storage.pushStateToHistory() in core after a search is completed. core.shouldPushState() was added 
    to prevent an extra state from being pushed on page load, from say a defaultInitialSearch or `query`
    url parameter.

There were also a few cases where the SDK was relying on globalStorage and persistentStorage
being separate and desynced, which had to be fixed
1. The SpellCheck component was deleting SKIP_SPELL_CHECK and QUERY_TRIGGER from persistentStorage
    in its onCreate(), while still expecting these values to be in globalStorage when the api call was made.
    This was removed. Now, these are deleted from storage at the end of a core.search() or core.verticalSearch()
2. FilterOptions had a listener to `this.name` in globalStorage, and in this listener, update persistentStorage. 
    Unifying the storages caused this to infinite loop in certain scenarios. Because the original purpose of the
    `this.name` listener was for back navigation, and new HISTORY_POP_STATE key + listener was added, which
    triggers when back or forward navigation occurs. 
3. FilterOptions, in its constructor, was pulling `this.name` from globalStorage in order to get the initial state of
    its selected options. However, FilterOptions, when different filter options were selected, it would only update
    persistentStorage but not globalStorage. This never worked properly with Facets, where it could cause a Facet
    checkbox to be displayed as selected even when it did not impact the search, or vice versa. This happens
    much more frequently after unifying the storages, because previously, the value of `this.name` in globalStorage
    would always just be whatever the initial state of the page was, typically null, and if it was null would not do anything.
    By unifying the storages, the value of `this.name` in globalStorage is now kept in sync in persistentStorage, which
    WAS being updated, causing the bug to happen much more frequently.
    To fix this, now, if a FilterOptions is part of a Facets component, it doesn't try to set anything in storage.
4. In a similar vein, FilterSearch had a listener for a storageKey it used to only be setting in persistentStorage, which would run a search. This listener was updated to use HISTORY_POP_STATE. It was a small change to update this listener to work for back navigation, so I included it in this PR.

As part of the HISTORY_POP_STATE listener (number 3 directly above), this commit also fixes a preexisting bug where, for a static filterbox with searchOnChange = true, back navigation would trigger additional searches and cause the page to flash.
This was fixed by changing updateListener() to updateListener(false, true), in the HISTORY_POP_STATE listener,
to block the extra searches. 

J=SLAP-1027
TEST=manual, auto

- added an acceptance test for static filters, including back navigation for filterbox (which was fixed this pr,
for “free”)
- added an acceptance test for filtersearch back navigation (also fixed this pr)

- added a jest test to make sure we don’t run extra queries on on back nav when a static filterbox
with searchOnChange is on the page
added an acceptance test for filtersearch back navigation
- ie11 does a very nasty thing of aggressively caching all ajax requests by default. These cached requests end up ignored by testcafe's RequestLogger. Added IE11NoCacheHook to deal with this, using a workaround found on the testcafe github (for a slightly different issue). Also added network logs to our browserstack automate

- check that only one state is pushed per search on a page with facets
- checked that in 1.7.1, when you have a static filterbox, it's expected that selecting a filter,
then hitting the back button causes multiple searches to be called
- check that navigation and no results links are updated with the current query correctly
- check that in 1.7.1 range and daterange filters dont work with backnavigation in any way
and also do not apply the filter properly on load
- check that in 1.7.1 sort options will appear visually to apply the state from the url, but not
actually impact the search
it also does not work with back nav in any way
- check that in 1.7.1, the filtersearch half of the geolocation filter (where you search for a filter using 
its searchbar) does not work with either back nav or on page load
on page load it will persist the query inside its searchbar but not actually apply the filter
- check that in 1.7.1, filtersearch will work on load (it will persist both the filter and the query
in its personal searchbar) but does not work with back navigation
a new bug was introduced, where if you selected a filtersearch filter say filter APPLE,
selected a different filtersearch filter, say filter BANANA, hit backnav, which would not do anything
to filtersearch (filter BANANA) would still be displayed on the page, and then refreshed the page,
filter APPLE would be applied when previously filter BANANA would be applied. Fixing this was
a relatively small change, so it was included in this PR.
- check that in 1.7.1, on universal, if you land on a page with no initial search, run a search, then hit backnav, it’s expected to see QA submission on the page
check that sessionTrackingEnabled is passed correctly in requests
- test that getCurrentStateUrlMerged() will include query params that are in the current url but not in persistentStorage. Tested by typing "window.history.replaceState(null, '', '?a=b')" into the console, then loading up a no results page and seeing that the links had a=b as an additional param
